Stable Children

pdf opened the pull request:
This PR aims to provide stable element children, to enable persistent `Component`s.

This is not final code, as I'm not really happy with some of this, however I believe it is at least correctly functional unless I've missed something, and I'd like to start getting some feedback.  I think we absolutely need to do something about testing here though - this got rather complicated.

Supercedes #132

pdf commented on markup.go:
```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
+}
+
+// MarkupIf returns nil if cond is false, otherwise it returns the given markup.
+func MarkupIf(cond bool, markup ...Applyer) Applyer {
```
I don't really like having this, but no better options came to mind.

slimsag replied:
I've thought about this for a bit, and haven't come up with any great options either. I'll keep thinking about it.

pdf commented on markup.go:
```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
+
+// List wraps a list of Markup, Component, or HTML which is individually
+// applied to an HTML element or text node.
+func List(m ...ComponentOrHTML) ChildList {
```
It's possible for a user to shoot themselves in the foot here due to the empty interface types, and get a panic at render time, like so:

```go
var blah []ComponentOrHTML
vecty.List(blah)
```
When they should instead do:
```go
var blah []ComponentOrHTML
vecty.List(blah...)
```

pdf replied:
We could alternatively just retain:
```go
type List []ComponentOrHTML
```
instead of `ChildList`, but it felt odd that `Markup` was a function, whereas `List` was a type, so they would require different syntax to use.

slimsag replied:
> We could alternatively just retain: ...

The reason `Markup` is a function is because the most common use case for `Markup` is something like this:

```
return elem.Div(
	vecty.Markup(
		vecty.Style("float", "right"),
	),
	...
}
```

So that users don't have to write:

```
return elem.Div(
	[]vecty.Markup{
		vecty.Style("float", "right"),
	},
	...
}
```

Which wouldn't really match the style of other things inside a component's `Render` method.

---

In contrast to the above, `List` is _not_ commonly used inline inside a `elem.Foobar()` function call. For example, I do not really expect users to write:

```
return elem.Div(
	vecty.List(
		vecty.Text("Hello"),
		vecty.Text("World"),
	),
}
```

Because they could instead just write that as:

```
return elem.Div(
	vecty.Text("Hello"),
	vecty.Text("World"),
}
```

What I would expect users to do is use `vecty.List` in place of a regular `[]ComponentOrHTML`, for example in code like https://github.com/pdf/vecty/blob/e71f4c2cda8a9fcd493e78d1034bb9befed0ab2e/example/todomvc/components/pageview.go#L179-L185, because writing `ComponentOrHTML` is tedious. IMHO for this use case, `vecty.List()` being a function seems more ambiguous as to what it really does.

---

Because of the above, I would say that we should just retain:

```
type List []ComponentOrHTML
```

Also, with the above, they can't shoot themselves in the foot with at least that specific example:

```
var blah []ComponentOrHTML
vecty.List(blah)
```

because that is now a type conversion. Of course, there are other various ways they could still shoot themselves in the foot, though.

pdf replied:
The same applies foot-shooting caveat applies to the Markup function though:

```go
var markup []Applyer
vecty.Markup(markup) // wrong
```

The awkwardness though was that if users have a children var of `[]ComponentOrHTML`, and a markup var of `[]Applyer`, the syntax to use them is different if we retain the old `List` type:

```go
var children []ComponentOrHTML
var markup []Applyer

vecty.Markup(markup...)
vecty.List(children)
```

I... think I'm okay with changing `List` though...

slimsag replied:
Let's ignore the foot-shooting portion: I agree it occurs in both cases.

---

I'm aware of the awkwardness you describe, but you're assuming that users will be converting their own `children` list via `vecty.List(children)`, why? i.e. with this change, you are saying that code previously written as:

```
var children vecty.List
children = append(children, foobar)
return Div(
    children,
)
```

should instead become:

```
var children []ComponentOrHTML
children = append(children, foobar)
return Div(
    vecty.List(children),
)
```

Sure, the above could just be `var children vecty.ChildList` -- but that is kind-of my point: I don't see in which scenario people would use this `vecty.List` helper function?

pdf replied:
ack, I'll change this in the next pass.

pdf commented:
There are also certainly render efficiency improvements that can be made once we get correctness nailed.

slimsag commented on markup.go:
```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
 
+// Markup wraps a list of Applyer which is individually
+// applied to an HTML element or text node.
+func Markup(m ...Applyer) MarkupList {
```
I think the return type here should just be `Applyer`.

I think the description should be simplified to just:

```
// Markup returns an applyer that applies all of the given markup.
```

pdf replied:
I specifically _don't_ want `Applyer` to be a valid type at the top level, that would mess up our ability to work with `nil` correctly.

pdf replied:
There is still an issue that I forgot to mention - swapping `nil` for a `MarkupList` between renders will cause us problems - I actually considered not exporting `MarkupList` to avoid this, but having unexported return types can be annoying.

slimsag replied:
Having an unexported type be returned sounds a bit hacky.

> I specifically don't want Applyer to be a valid type at the top level, that would mess up our ability to work with nil correctly.

I'd really like to understand this more, could you point me to relevant code where this being an `Applyer` would cause trouble? Or provide a small example demo'ing the problem?

pdf replied:
`nil` now occupies a child slot, so anything that returned `nil` in place of `Applyer` would break component persistence.

If `Applyer` is valid at the top level (in which case, there's no point having a `Markup` function, and `Applyer` could just stay named `Markup`):
```go
func (c *MyComponent) OptionalMarkup() vecty.Applyer {
        if c.condition {
                return vecty.Property("key", "value")
        }

        return nil
}

func (c *MyComponent) Render() *vecty.HTML {
        return elem.Div(
                c.OptionalMarkup(), // breaks persistence
                &OtherComponent{},
        )
}
```

We currently have a similar problem with `MarkupList`:
```go
func (c *MyComponent) OptionalMarkup() vecty.MarkupList {
        if c.condition {
                return vecty.Markup(vecty.Property("key", "value"))
        }

        return nil
}

func (c *MyComponent) Render() *vecty.HTML {
        return elem.Div(
                c.OptionalMarkup(), // breaks persistence
                &OtherComponent{},
        )
}
```

However, that becomes impossible if `MarkupList` is unexported, since it cannot be used as a return value, so we will never be able to receive a concrete `nil` in it's place, and it forces the use of `vecty.Markup()` for every case.

slimsag replied:
Ahhh I see, understood. I had forgotten that `MarkupList` must effectively sit side-by-side with `Component` and `*HTML`, and that nil `*HTML` occupies a child slot.

This would not be the case of we forced users to supply Markup always:

```
func Div(markup vecty.Markup, children ...vecty.ComponentOrHTML) *vecty.HTML {
    ...
}
```

But understandably that would be a worse API for situations like `elem.Paragraph(vecty.Text("hello world!"))`.

I understand why you proposed an unexported `MarkupList` above, now, and agree that sounds best.

slimsag commented on markup.go:
```diff
@@ -154,26 +150,49 @@ func (m ClassMap) Apply(h *HTML) {
-	for _, m := range l {
-		apply(m, h)
+// Apply implements the Applyer interface.
+func (m MarkupList) Apply(h *HTML) {
```
I personally wonder if having this exported makes sense or not.

I think the number of situations where someone will use it is quite low, and they could use `var markup []vecty.Applyer` & `vecty.Markup(markup...)` for such cases.

My instinct says this should be unexported until we have good evidence that it would be fairly useful to people, this way we keep our exported API footprint smaller.

pdf replied:
Yep, this was left exposed by accident.

slimsag commented on markup.go:
```diff
@@ -47,36 +47,32 @@ type Event struct {
-	case Component:
-		h.children = append(h.children, m)
-	case *HTML:
+	case MarkupList:
```
This case would just become `Applyer`

pdf replied:
See https://github.com/gopherjs/vecty/pull/134#discussion_r135370011

slimsag commented on markup.go:
```diff
@@ -47,36 +47,32 @@ type Event struct {
+//  MarkupList
 //  Component
 //  *HTML
+//  ChildList
```
`nil` should also be added here.

pdf replied:
Yeah, I noticed this after I left my other comments, but didn't want to force-push just for that.

slimsag commented on markup.go:
```diff
@@ -47,36 +47,32 @@ type Event struct {
 //
 // If the underlying value is not one of these types, the code handling the
 // value is expected to panic.
 type MarkupOrComponentOrHTML interface{}
```
The name now does not reflect the actual type. But of course `MarkupListOrComponentOrHTMLOrChildListOrNil` is not a great name.

Perhaps shorter more abstract name would do us good, something like `MarkupOrChild`.

pdf replied:
Sounds fair to me.

slimsag commented on markup.go:
```diff
@@ -47,36 +47,32 @@ type Event struct {
 // MarkupOrComponentOrHTML represents one of:
 //
-//  Markup
+//  MarkupList
```
Couldn't this be `Applyer` instead?

pdf replied:
See https://github.com/gopherjs/vecty/pull/134#discussion_r135370011

slimsag commented on dom.go:
```diff
@@ -347,6 +483,15 @@ func extractHTML(e ComponentOrHTML) *HTML {
+		return true
+	default:
+		return false
+	}
```
Please use simpler syntax:

```
_, ok := e.(ChildList)
return ok
```

pdf replied:
ack

slimsag commented on dom.go:
```diff
@@ -245,22 +247,69 @@ func (h *HTML) reconcile(prev *HTML) {
+func (h *HTML) reconcileChildren(prev *HTML, insertBefore *HTML) {
+	if prev == nil {
+		prev = &HTML{}
+	}
```
What does this do?

pdf replied:
Allows calling `reconcileChildren` on the first pass where the prev render is `nil`, sames as early in `*HTML.reconcile()`.

pdf replied:
This may actually be unnecessary now, as some of the calling code has changed, will review.

slimsag commented on dom.go:
```diff
@@ -245,22 +247,69 @@ func (h *HTML) reconcile(prev *HTML) {
+			} else {
+				h.node.Call("appendChild", nextChildRender.node)
 				continue
 			}
```
Prefer less indention, 273 already has `continue` so just remove `else`

pdf replied:
ack

slimsag commented on dom.go:
```diff
@@ -276,18 +325,105 @@ func (h *HTML) reconcile(prev *HTML) {
+	nextHTML := &HTML{node: node}
+	nextHTML.removeChildren(&HTML{node: node, children: children})
+}
+
```
All of the code in this file above this line is quite complex, and I don't understand most of it currently (I've only had ~15 mins to look at it, my lunch break is up now).

I'll have more time to look at this later today or tomorrow, but I think we should be able to simplify a lot of this or make it more clear what is going on. Hopefully my other comments will at least let us solidify the public API.

pdf replied:
Thanks @slimsag, open to suggestions - as I said in the OP, I do not expect this to be final code, but a first functional pass.  The list handling is necessarily somewhat complex though.

slimsag replied:
Understandable :) I'm going to take a more deep look at this code tonight or tomorrow and offer more concrete suggestions than just "the code is complex" :P I think this PR is a really good starting point!

pdf commented:
I have the current review notes queued (with some other minor style changes), but I'll wait for comments on the list handling before doing a rebased push.

slimsag commented:
@pdf feel free to force push now. It'd be good for me to get a view of it with all pending comments fixed.

pdf commented:
`golint` doesn't like the unexported return value.

pdf commented:
I think we have two options:
1. keep the unexported return and suppress the lint warning as I've done here, by moving to `gometalinter` (not a terrible idea anyway, as we get some extra checks there)
2. use two arguments to separate markup from children, though this makes the API awkward

slimsag commented on markup.go:
```diff
@@ -154,26 +151,43 @@ func (m ClassMap) Apply(h *HTML) {
 
-// UnsafeHTML is Markup which unsafely sets the inner HTML of an HTML element.
+// MarkupIf returns nil if cond is false, otherwise it returns the given markup.
+func MarkupIf(cond bool, markup ...Applyer) Applyer {
```
Still an open question to me if there is a better API than this. But I won't block this PR on that, it's something we can improve in the future if we find a better alternative.

slimsag commented on .travis.yml:
```diff
@@ -6,20 +6,19 @@ install:
   - npm install -g source-map-support
   - go get -u github.com/gopherjs/gopherjs
-  - go get -u github.com/golang/lint/golint
+  - go get -u gopkg.in/alecthomas/gometalinter.v1
```
I've just realized a second benefit to this approach. Imagine a situation where a user has some very complex markup for many different logical cases (e.g. themes or something like that) and they want to split it out into separate functions:

```Go
func GetTheme(name string) []Applyer {
	...
}
```

Callers would have to use it like this:

```Go
return elem.Div(
	vecty.Markup(foo.GetTheme("dark")...),
	...
)
```

Which, if you're a complete beginner, does look a bit strange. This is because `GetTheme` can't return the actual value of `vecty.Markup` to the user of the package. In contrast, with it being exported as suggested above, they could write:

```Go
func GetTheme(name string) vecty.MarkupList {
	return vecty.Markup(getTheme("dark")...)
}
```

And then callers may just use it like:

```Go
return elem.Div(
	foo.GetTheme("dark"),
	...
)
```

pdf replied:
Just to clarify a few things about `gometalinter`.  It's been around for a long time, and is pretty widely used.

1. It _vendors_ checks so that versions are stable, and installs always succeed, but they've had plenty of changes upstreamed to those projects.  megacheck is able to get away with not vendoring since all the checks are written by the same author.
2. Honestly, I'm not sure whether new linters will be added to v1 or not, but being explicit about what checks are run seems fine to me either way, otherwise new checks may unexpectedly break your CI.
3. The config file is optional, I just used it to make the CI config clearer, since we special-case `golint` for the generated code.  Choosing which checks to run is already something every project does, whether the checks are bundled under a single runner or not, by choosing what to enable in the CI run - the box is already open ;-)
4. Not exactly, it just makes it easy to run multiple existing checks via a single executable call - this is particularly useful for editor integration.

megacheck is very new by comparison (though I'm sure the quality is high), wrapping some of Dominik's other tools, though they've been available via gometalinter for a long time.  I'm not really phased either way, but I thought some of the criticisms were a little hard on gometalinter.

---

As for the API suggestion here, I like it :+1:

slimsag commented:
Sorry for the delay on reviewing this @pdf -- took me longer to get around to this than I thought it would.

I know today (Sun for me) is a weekday for you, so in addition to the (many) inline comments I've made above, I have also addressed each and every one of them and sent it as a PR to your branch here via https://github.com/pdf/vecty/pull/1

If I assume that you mostly agree with the comments I've made above, then I see three remaining issues for this PR before merging:

1. This comment https://github.com/gopherjs/vecty/pull/134#discussion_r135441646
2. Testing
3. Updating `docs/CHANGELOG.md`

On the topic of testing, I have been thinking about it a fair bit. I have some good ideas on how to handle this, by essentially creating some nice helpers in our test suite and then describing results effectively just as a list of HTML / Component operations that occurred. It will probably be a few more weeks before I can get to this, however, and I feel that this PR is a clear improvement over the current status quo (pending resolution of the comment #1 mentioned above) so I'd be fine merging this PR without tests once you feel confident about this being the right approach as well.

slimsag commented:
Will need to update the CI to use Go 1.9 as well since it is now failing due to GopherJS requiring latest Go version.

pdf commented on dom.go:
```diff
@@ -276,22 +321,109 @@ func (h *HTML) reconcile(prev *HTML) {
+		// better method for handling this.
+		idx, _ := l[exceptIndex].(List).firstHTML()
+		l[exceptIndex].(List).removeExcept(node, idx)
+		children = append(l[:exceptIndex], l[exceptIndex+1:]...)
```
When we render `List`s, we do so recursively - all elements of all nested lists end up rendered as children of the first outer list's parent node, so when operating in the reverse we need to ensure that we recurs all child lists there too.

I wrote this logic before I was certain whether we would support a `nil` child though, so before final merge, I think I may just remove `exceptIndex` altogether and recursively remove all list elements, then we'll append/insert the new child.  It's an extra DOM operation, but I think that's okay for clarity.  Will need to do some testing though.

pdf commented:
@slimsag no problem, thanks for the review - the changes look good to me.

I'm very keen to get this tested, but improvements to the test harness would make the job significantly easier.  Though this is a significant breaking change, I think once it's finalised this should go in regardless, as the current code on master is prone to unexpected behaviour, and I think this is a sensible approach to resolving that.

pdf commented on dom.go:
```diff
@@ -276,22 +321,109 @@ func (h *HTML) reconcile(prev *HTML) {
+	if prev == nil {
+		return
 	}
 	for i := len(h.children); i < len(prev.children); i++ {
```
Because in the case where there are no prev.children, this will panic, I've omitted this change.

slimsag replied:
Got it, I missed that. Thanks for omitting!

pdf commented:
I merged your changes with a couple of minor modifications, added the change mentioned in https://github.com/gopherjs/vecty/pull/134#discussion_r135463934, and cleaned up the history.  I think this is good for final review now.

I tried bumping the Go version on Travis, but looks like they haven't finished rolling out 1.9 support yet.

slimsag commented:
Try tip version for Travis? May work.

shurcooL commented:
`go1.9.x` syntax seems to be working in Travis now. See https://travis-ci.org/go-interpreter/wagon/jobs/269323333.

pdf commented:
Thanks for the heads-up @shurcooL, looks like we still need the `GOROOT` hack though.

pdf commented:
And travis has fixed the GOROOT issue now too, no more 1.9 hacks required.

slimsag commented:
> Though this is a significant breaking change, I think once it's finalised this should go in regardless, as the current code on master is prone to unexpected behaviour, and I think this is a sensible approach to resolving that.

I agree highly.

I'll do one final look through of this change + write up the doc/CHANGELOG.md as soon as I can, possibly tonight (Thur) or Saturday morning.

slimsag commented:
Also, do you have any final concerns about this PR? Feel it's good to go in its current state?

pdf commented:
It occurred to me the other night that we could now get rid of the `<noscript>` output for `nil` returns from component renders.  I'd like to include that too, and it shouldn't be too much effort, though I might not finish it this morning before work.

pdf commented:
Actually, allowing `nil` renders from components will have some implications for `vecty.Rerender()` that I may not have time to resolve this weekend, so with a couple of changes to support that approach at a later, I'm going to call this PR now, and we can address `<noscript>` in a subsequent PR.

slimsag commented:
Sounds good.

On Aug 31, 2017 4:40 PM, "Peter Fern" <notifications@github.com> wrote:

> Actually, allowing nil renders from components will have some
> implications for vecty.Rerender() that I may not have time to resolve
> this weekend, so with a couple of changes to support that approach at a
> later, I'm going to call this PR now, and we can address <noscript> in a
> subsequent PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gopherjs/vecty/pull/134#issuecomment-326448630>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADBrOF3Yo_FCyg5pLo1P21KapvEnMANxks5sd0RugaJpZM4PCrOK>
> .
>

slimsag commented:
Alright, I've just now had time to do a final pass over this PR. Nothing sticks out to me, so I am going to go ahead and merge this and update the `doc/CHANGELOG.md` directly following.

slimsag commented:
Changelog updated, see https://github.com/gopherjs/vecty/blob/master/doc/CHANGELOG.md#sept-2-2017-pr-134-major-breaking-change

pdf commented:
Changelog looks good.  Is it worth making mention on slack or somewhere of the breakage?

slimsag commented:
Yep :) I mentioned in Slack channel as well
